File: src-tauri/src/models/music.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Artist {
    pub name: String,
    pub id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Album {
    pub name: String,
    pub id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SongItem {
    pub id: String,
    pub title: String,
    pub artists: Vec<Artist>,
    pub album: Option<Album>,
    pub duration: Option<u64>,
    pub music_video_type: Option<String>,
    pub thumbnail: String,
    pub explicit: bool,
    pub video_id: Option<String>,
    pub playlist_id: Option<String>,
    pub params: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AlbumItem {
    pub browse_id: String,
    pub playlist_id: String,
    pub title: String,
    pub artists: Option<Vec<Artist>>,
    pub year: Option<i32>,
    pub thumbnail: String,
    pub explicit: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaylistItem {
    pub id: String,
    pub title: String,
    pub author: Option<Artist>,
    pub song_count_text: Option<String>,
    pub thumbnail: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArtistItem {
    pub id: String,
    pub title: String,
    pub thumbnail: Option<String>,
    pub channel_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EpisodeItem {
    pub id: String,
    pub title: String,
    pub author: Option<Artist>,
    pub thumbnail: String,
    pub explicit: bool,
    pub publish_date_text: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PodcastItem {
    pub id: String,
    pub title: String,
    pub author: Option<Artist>,
    pub episode_count_text: Option<String>,
    pub thumbnail: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum YTItem {
    Song(SongItem),
    Album(AlbumItem),
    Playlist(PlaylistItem),
    Artist(ArtistItem),
    Episode(EpisodeItem),
    Podcast(PodcastItem),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArtistSection {
    pub title: String,
    pub items: Vec<YTItem>,
    pub more_endpoint_browse_id: Option<String>,
    pub more_endpoint_params: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArtistPage {
    pub artist: ArtistItem,
    pub sections: Vec<ArtistSection>,
    pub description: Option<String>,
    pub subscriber_count_text: Option<String>,
    pub monthly_listener_count: Option<String>,
    pub is_subscribed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MoodAndGenreItem {
    pub title: String,
    pub stripe_color: u64,
    pub browse_id: String,
    pub params: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExplorePage {
    pub new_release_albums: Vec<AlbumItem>,
    pub mood_and_genres: Vec<MoodAndGenreItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChartSection {
    pub title: String,
    pub items: Vec<YTItem>,
    pub chart_type: String, // "Trending" | "Top" | "Genre" | "NewReleases"
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChartsPage {
    pub sections: Vec<ChartSection>,
    pub continuation: Option<String>,
}

File: src/types/music.ts
// TypeScript mirror of the Rust YouTube Music models
// (`src-tauri/src/models/music*.rs`). All keys are camelCase to match serde.

export interface Artist {
  name: string;
  id: string | null;
}

export interface Album {
  name: string;
  id: string;
}

export interface SongItem {
  id: string;
  title: string;
  artists: Artist[];
  album: Album | null;
  duration: number | null;
  musicVideoType: string | null;
  thumbnail: string;
  explicit: boolean;
  videoId: string | null;
  playlistId: string | null;
  params: string | null;
}

export interface AlbumItem {
  browseId: string;
  playlistId: string;
  title: string;
  artists: Artist[] | null;
  year: number | null;
  thumbnail: string;
  explicit: boolean;
}

export interface PlaylistItem {
  id: string;
  title: string;
  author: Artist | null;
  songCountText: string | null;
  thumbnail: string | null;
}

export interface ArtistItem {
  id: string;
  title: string;
  thumbnail: string | null;
  channelId: string | null;
}

export interface EpisodeItem {
  id: string;
  title: string;
  author: Artist | null;
  thumbnail: string;
  explicit: boolean;
  publishDateText: string | null;
}

export interface PodcastItem {
  id: string;
  title: string;
  author: Artist | null;
  episodeCountText: string | null;
  thumbnail: string | null;
}

// Internally-tagged enum (`#[serde(tag = "type")]`) — discriminated union.
export type YTItem =
  | ({ type: "song" } & SongItem)
  | ({ type: "album" } & AlbumItem)
  | ({ type: "playlist" } & PlaylistItem)
  | ({ type: "artist" } & ArtistItem)
  | ({ type: "episode" } & EpisodeItem)
  | ({ type: "podcast" } & PodcastItem);

// --- Sections / pages -----------------------------------------------------

export interface ArtistSection {
  title: string;
  items: YTItem[];
  moreEndpointBrowseId: string | null;
  moreEndpointParams: string | null;
}

export interface ArtistPage {
  artist: ArtistItem;
  sections: ArtistSection[];
  description: string | null;
  subscriberCountText: string | null;
  monthlyListenerCount: string | null;
  isSubscribed: boolean;
}

export interface MoodAndGenreItem {
  title: string;
  stripeColor: number;
  browseId: string;
  params: string | null;
}

export interface ExplorePage {
  newReleaseAlbums: AlbumItem[];
  moodAndGenres: MoodAndGenreItem[];
}

export interface ChartSection {
  title: string;
  items: YTItem[];
  chartType: string; // "Trending" | "Top" | "Genre" | "NewReleases"
}

export interface ChartsPage {
  sections: ChartSection[];
  continuation: string | null;
}

export interface MusicHomeChip {
  title: string;
  browseId: string | null;
  params: string | null;
  orderBy: number;
}

export interface MusicShelf {
  title: string;
  subtitle: string | null;
  browseId: string | null;
  params: string | null;
  items: YTItem[];
}

export interface MusicHomePage {
  chips: MusicHomeChip[];
  sections: MusicShelf[];
  continuation: string | null;
}

export interface MusicSearchSection {
  title: string;
  items: YTItem[];
}

export interface MusicSearchResponse {
  sections: MusicSearchSection[];
  continuation: string | null;
}

export interface SearchSummaryPage {
  summaries: MusicSearchSection[];
}

export interface MusicSearchSuggestions {
  queries: string[];
  recommendedItems: YTItem[];
}

export interface AlbumPage {
  album: AlbumItem;
  description: string | null;
  songCount: number | null;
  durationText: string | null;
  songs: SongItem[];
  continuation: string | null;
}

export interface MusicPlaylistPage {
  id: string;
  title: string;
  author: Artist | null;
  songCountText: string | null;
  thumbnail: string | null;
  description: string | null;
  songs: SongItem[];
  continuation: string | null;
}

export interface RelatedPage {
  songs: SongItem[];
  albums: AlbumItem[];
  artists: ArtistItem[];
  playlists: PlaylistItem[];
}

/** A Daily Mix cluster from the music brain: a label + seed tracks to expand. */
export interface DailyMixSeed {
  label: string;
  seedTrackIds: string[];
}

export interface QueuePage {
  items: SongItem[];
  currentIndex: number | null;
  continuation: string | null;
  lyricsBrowseId: string | null;
  lyricsParams: string | null;
  relatedBrowseId: string | null;
  radioPlaylistId: string | null;
}

export interface MoodGenrePage {
  title: string;
  items: YTItem[];
  continuation: string | null;
}

export interface MusicStreamInfo {
  videoId: string;
  audioUrl: string;
  mimeType: string;
  itag: number;
  bitrate: number | null;
  approxDurationMs: number | null;
  loudnessDb: number | null;
  perceptualLoudnessDb: number | null;
  expiresInSeconds: number;
  usedClient: string;
}

/// `(songs, continuation)` tuple returned by album/playlist continuation commands.
export type SongContinuation = [SongItem[], string | null];
